applications/serial_lte_modem/src/slm_at_commands.c

```diff
@@ -0,0 +1,580 @@
+int handle_at_recvfrom(enum at_cmd_type cmd_type);
+int handle_at_getaddrinfo(enum at_cmd_type cmd_type);
+
+#if defined(CONFIG_SLM_NATIVE_TLS)
+int handle_at_xcmng(enum at_cmd_type cmd_type);
+#endif
+
+/* ICMP commands */
+int handle_at_icmp_ping(enum at_cmd_type cmd_type);
+
+/* FOTA commands */
+int handle_at_fota(enum at_cmd_type cmd_type);
+
+#if defined(CONFIG_SLM_GPS)
+int handle_at_gps(enum at_cmd_type cmd_type);
+#endif
+
+#if defined(CONFIG_SLM_FTPC)
+int handle_at_ftp(enum at_cmd_type cmd_type);
+#endif
+
+#if defined(CONFIG_SLM_MQTTC)
+int handle_at_mqtt_connect(enum at_cmd_type cmd_type);
+int handle_at_mqtt_publish(enum at_cmd_type cmd_type);
+int handle_at_mqtt_subscribe(enum at_cmd_type cmd_type);
+int handle_at_mqtt_unsubscribe(enum at_cmd_type cmd_type);
+#endif
+
+#if defined(CONFIG_SLM_HTTPC)
+int handle_AT_HTTPC_CONNECT(enum at_cmd_type cmd_type);
+int handle_AT_HTTPC_REQUEST(enum at_cmd_type cmd_type);
+#endif
+
+static slm_at_cmd_t slm_at_cmd_list[] = {
+	/* Generic commands */
+	{"AT#XSLMVER", handle_at_slmver},
+	{"AT#XSLEEP", handle_at_sleep},
+	{"AT#XRESET", handle_at_reset},
+	{"AT#XCLAC", handle_at_clac},
+	{"AT#XSLMUART", handle_at_slmuart},
+	{"AT#XDATACTRL", handle_at_datactrl},
+
+	/* TCP proxy commands */
+	{"AT#XTCPFILTER", handle_at_tcp_filter},
+	{"AT#XTCPSVR", handle_at_tcp_server},
+	{"AT#XTCPCLI", handle_at_tcp_client},
+	{"AT#XTCPSEND", handle_at_tcp_send},
+	{"AT#XTCPRECV", handle_at_tcp_recv},
+
+	/* UDP proxy commands */
+	{"AT#XUDPSVR", handle_at_udp_server},
+	{"AT#XUDPCLI", handle_at_udp_client},
+	{"AT#XUDPSEND", handle_at_udp_send},
+
+	/* Socket-type TCPIP commands */
+	{"AT#XSOCKET", handle_at_socket},
+	{"AT#XSOCKETOPT", handle_at_socketopt},
+	{"AT#XBIND", handle_at_bind},
+	{"AT#XCONNECT", handle_at_connect},
+	{"AT#XLISTEN", handle_at_listen},
+	{"AT#XACCEPT", handle_at_accept},
+	{"AT#XSEND", handle_at_send},
+	{"AT#XRECV", handle_at_recv},
+	{"AT#XSENDTO", handle_at_sendto},
+	{"AT#XRECVFROM", handle_at_recvfrom},
+	{"AT#XGETADDRINFO", handle_at_getaddrinfo},
+
+#if defined(CONFIG_SLM_NATIVE_TLS)
+	/* Hijacked modem command */
+	{"AT%CMNG", handle_at_xcmng},
+#endif
+	/* ICMP commands */
+	{"AT#XPING", handle_at_icmp_ping},
+
+	/* FOTA commands */
+	{"AT#XFOTA", handle_at_fota},
+
+#if defined(CONFIG_SLM_GPS)
+	/* GPS commands */
+	{"AT#XGPS", handle_at_gps},
+#endif
+
+#if defined(CONFIG_SLM_FTPC)
+	/* FTP commands */
+	{"AT#XFTP", handle_at_ftp},
+#endif
+
+#if defined(CONFIG_SLM_MQTTC)
+	{"AT#XMQTTCON", handle_at_mqtt_connect},
+	{"AT#XMQTTPUB", handle_at_mqtt_publish},
+	{"AT#XMQTTSUB", handle_at_mqtt_subscribe},
+	{"AT#XMQTTUNSUB", handle_at_mqtt_unsubscribe},
+#endif
+
+#if defined(CONFIG_SLM_HTTPC)
+	{"AT#XHTTPCCON", handle_AT_HTTPC_CONNECT},
+	{"AT#XHTTPCREQ", handle_AT_HTTPC_REQUEST},
+#endif
+};
+
+int handle_at_clac(enum at_cmd_type cmd_type)
+{
+	int ret = -EINVAL;
+
+	if (cmd_type == AT_CMD_TYPE_SET_COMMAND) {
+		int total = sizeof(slm_at_cmd_list)/sizeof(slm_at_cmd_t);
```

Comment:
You could use `ARRAY_SIZE(slm_at_cmd_list)` utility macro to obtain number of array elements.

Reply:
Taken with gratitude